python_modules/dagster/dagster/_core/definitions/utils.py

```diff
@@ -393,3 +395,17 @@ def config_from_pkg_resources(pkg_resource_defs: Sequence[Tuple[str, str]]) -> M
         ) from err
 
     return config_from_yaml_strings(yaml_strings=yaml_strings)
+
+
+T = TypeVar("T")
+
+
+def dedupe_objects_ref_equality(objects: Iterable[T]) -> Sequence[T]:
+    """Dedupe definitions by reference equality."""
+    seen = set()
+    deduped = []
+    for obj in objects:
+        if obj not in seen:
+            deduped.append(obj)
+            seen.add(obj)
+    return deduped
```

Comment:
I don't think this is actual reference equality. We're using a set, which hashes the object. This is hash code equality.

Using `is` would test for reference equality if we wanted that.

Also, we can simplify this if we choose to continue with hash code equality.

```suggestion
def deduplicate_by_hashcode_for_objects(objects: Iterable[T]) -> Sequence[T]:
    """Dedupe definitions by reference equality."""
    return list(set(objects))
```

Reply:
I think set equality is both hash code _and_ object equality right?

Reply:
but agreed with your simplification.

Reply:
> I think set equality is both hash code _and_ object equality right? https://stackoverflow.com/questions/63544290/what-does-set-in-python-use-to-test-equality-between-objects

Yeah, you're right.

Reply:
Yeah sets invoke `__hash__` and then `__eq__` on collisions.

Default implementations of those look something like this (written in C):

```python
class A:
    def __hash__(self) -> int:
        return id(self)

    def __eq__(self, other: object) -> bool:
        return id(self) == id(other)
```

For fun I asked ChatGPT to show us what the underling CPython [implementation](https://chatgpt.com/share/6900f479-1ac0-4e53-94a6-492806a25cb2) looks like.